pkg/operator/controller/controller.go

```diff
@@ -302,11 +374,83 @@ func (r *ReconcileCDI) reconcileUpdate(logger logr.Logger, cr *cdiv1alpha1.CDI)
 
 	if ready {
 		logger.Info("Operator is ready!!")
+		if r.isUpgrading(cr) {
+			if err = r.cleanupUnusedResources(logger, cr); err != nil {
+				return reconcile.Result{}, err
+			}
+			previousVersion := cr.Status.ObservedVersion
+			cr.Status.ObservedVersion = r.namespacedArgs.DockerTag
+			cr.Status.OperatorVersion = r.namespacedArgs.DockerTag
+
+			//Is there a possible race if cr is marked as deleted during upgrade?
+			//we want to set cr in DeployedPhase, but may be it is being marked as Deleted already
+			if err = r.crUpdate(cdiv1alpha1.CDIPhaseDeployed, cr); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			logger.Info("Successfully finished Upgrade from %s to %s and entered Deployed state", previousVersion, cr.Status.ObservedVersion)
+		}
 	}
 
 	return reconcile.Result{}, nil
 }
 
+func (r *ReconcileCDI) cleanupUnusedResources(logger logr.Logger, cr *cdiv1alpha1.CDI) error {
+	//Iterate over installed resources of
+	//Deployment/CRDs/Services etc and delete all resources that
+	//do not exist in current version
+
+	targetStrategy, err := r.getAllResources(cr)
+	if err != nil {
+		return err
+	}
+
+	listTypes := []runtime.Object{
+		&extv1beta1.CustomResourceDefinitionList{},
+		&appsv1.DeploymentList{},
+		&corev1.ServiceList{},
+		&rbacv1.ClusterRoleBindingList{},
+		&rbacv1.ClusterRoleList{},
+		&rbacv1.RoleBindingList{},
+		&rbacv1.RoleList{},
+		&corev1.ServiceAccountList{},
+	}
+
+	for _, lt := range listTypes {
+		lo := &client.ListOptions{}
+		lo.SetLabelSelector("cdi.kubevirt.io")
+
+		if err := r.client.List(context.TODO(), lo, lt); err != nil {
+			logger.Error(err, "Error listing resources")
+			return err
+		}
+
+		sv := reflect.ValueOf(lt).Elem()
+		iv := sv.FieldByName("Items")
+
+		for i := 0; i < iv.Len(); i++ {
+			obj := iv.Index(i).Addr().Interface().(runtime.Object)
+			found := false
+			for _, target := range targetStrategy {
+				if reflect.TypeOf(obj) == reflect.TypeOf(target) {
+					if obj.(metav1.Object).GetName() == target.(metav1.Object).GetName() {
```

Comment:
Should we also check namespace just to be safe?

Reply:
done